Fix unexpected failure of CalculateRoot

This was noticed on stressnet while running TPS test.  While the fix is pretty simple, it's unclear what has suddenly changed 
to cause this panic.

```
ERROR: 2021/06/15 13:41:39 grpc: server failed to encode response: rpc error: code = Internal desc = grpc: error while marshaling: proto: Marshal called with nil
panic: interface conversion: interface is nil, not merkletree.Payload

goroutine 3167264 [running]:
  dusk-blockchain/pkg/core/data/block.(*Block).CalculateRoot(0xc008e49a60, 0xc008e49a60, 0x21, 0xc00eedf2f0, 0xc005710000, 0x17a)
           dusknetwork/dusk-blockchain/pkg/core/data/block/block.go:61 +0xb1
  dusk-blockchain/pkg/core/consensus/blockgenerator/candidate.(*generator).GenerateBlock(0xc01adf2ee0, 0xb7, 0xc008a12880, 0x21, 0x40, 0xc00fe86740, 0x20, 0x20, 0xc0085315a0, 0x1, ...)
           dusknetwork/dusk-blockchain/pkg/core/consensus/blockgenerator/candidate/blockgenerator.go:139 +0x2cd
  dusk-blockchain/pkg/core/consensus/blockgenerator/candidate.(*generator).Generate(...)
           dusknetwork/dusk-blockchain/pkg/core/consensus/blockgenerator/candidate/blockgenerator.go:110
  dusk-blockchain/pkg/core/consensus/blockgenerator/candidate.(*generator).GenerateCandidateMessage(0xc01adf2ee0, 0x1056ec0, 0xc00366f780, 0xb7, 0xc00791c190, 0xa, 0xa, 0xc001e0e600, 0xc0024686c0, 0x21, ...)
           dusknetwork/dusk-blockchain/pkg/core/consensus/blockgenerator/candidate/blockgenerator.go:79 +0x37a
  dusk-blockchain/pkg/core/consensus/selection.(*Phase).Run(0xc000112ee0, 0x1056ec0, 0xc00366f700, 0xc0007866e0, 0xc00028c4e0, 0xb7, 0xc00791c190, 0xa, 0xa, 0xc001e0e600, ...)
           dusknetwork/dusk-blockchain/pkg/core/consensus/selection/step.go:103 +0xca3
  dusk-blockchain/pkg/core/loop.(*Consensus).Spin(0xc000046e00, 0x1056ec0, 0xc00366f6c0, 0x1040140, 0xc000112ee0, 0x10400a0, 0xc008531360, 0xb7, 0xc00791c190, 0xa, ...)
           dusknetwork/dusk-blockchain/pkg/core/loop/consensus.go:157 +0x29c
  dusk-blockchain/pkg/core/chain.(*Chain).produceBlock.func1(0xc00f3f4420, 0xc000132820, 0x1056ec0, 0xc00366f6c0, 0x1040140, 0xc000112ee0, 0x10400a0, 0xc008531360, 0xb7, 0xc00791c190, ...)
           dusknetwork/dusk-blockchain/pkg/core/chain/chain.go:266 +0xb8
created by   dusk-blockchain/pkg/core/chain.(*Chain).produceBlock
           dusknetwork/dusk-blockchain/pkg/core/chain/chain.go:265 +0x37c

```

--- a/cmd/dusk/server.go
+++ b/cmd/dusk/server.go
@@ -239,7 +239,7 @@ func Setup() *Server {
 
 	// Create the listener and contact the voucher seeder
 	var gossip *protocol.Gossip
-	if cfg.Get().Kadcast.Enabled {
+	if !cfg.Get().Kadcast.Enabled {
 		gossip = protocol.NewGossip(protocol.TestNet)
 		connector := peer.NewConnector(eventBus, gossip, cfg.Get().Network.Port, processor, protocol.ServiceFlag(cfg.Get().Network.ServiceFlag), peer.Create)
 

--- a/cmd/dusk/server.go
+++ b/cmd/dusk/server.go
@@ -238,18 +238,21 @@ func Setup() *Server {
 	readerFactory := peer.NewReaderFactory(processor)
 
 	// Create the listener and contact the voucher seeder
-	gossip := protocol.NewGossip(protocol.TestNet)
-	connector := peer.NewConnector(eventBus, gossip, cfg.Get().Network.Port, processor, protocol.ServiceFlag(cfg.Get().Network.ServiceFlag), peer.Create)
+	var gossip *protocol.Gossip
+	if cfg.Get().Kadcast.Enabled {
+		gossip = protocol.NewGossip(protocol.TestNet)
+		connector := peer.NewConnector(eventBus, gossip, cfg.Get().Network.Port, processor, protocol.ServiceFlag(cfg.Get().Network.ServiceFlag), peer.Create)
 
-	seeders := cfg.Get().Network.Seeder.Addresses
-	for _, seeder := range seeders {
-		if err = connector.Connect(seeder); err != nil {
-			log.WithError(err).Error("could not contact voucher seeder")
+		seeders := cfg.Get().Network.Seeder.Addresses
+		for _, seeder := range seeders {
+			if err = connector.Connect(seeder); err != nil {
+				log.WithError(err).Error("could not contact voucher seeder")
+			}
 		}
-	}
 
-	if connector.GetConnectionsCount() == 0 {
-		panic("could not contact any voucher seeders")
+		if connector.GetConnectionsCount() == 0 {
+			panic("could not contact any voucher seeders")
+		}
 	}
 
 	// creating the Server

--- a/cmd/dusk/server.go
+++ b/cmd/dusk/server.go
@@ -238,9 +238,9 @@ func Setup() *Server {
 	readerFactory := peer.NewReaderFactory(processor)
 
 	// Create the listener and contact the voucher seeder
-	var gossip *protocol.Gossip
+	gossip := protocol.NewGossip(protocol.TestNet)
+
 	if !cfg.Get().Kadcast.Enabled {
-		gossip = protocol.NewGossip(protocol.TestNet)
 		connector := peer.NewConnector(eventBus, gossip, cfg.Get().Network.Port, processor, protocol.ServiceFlag(cfg.Get().Network.ServiceFlag), peer.Create)
 
 		seeders := cfg.Get().Network.Seeder.Addresses

--- a/harness/engine/network.go
+++ b/harness/engine/network.go
@@ -92,12 +92,21 @@ func (s *sessionlessClient) GracefulClose(options ...grpc.DialOption) {
 	}
 }
 
+const (
+	// GossipNetwork Gossip network type.
+	GossipNetwork = byte(0)
+	// KadcastNetwork  Kadcast network type.
+	KadcastNetwork = byte(1)
+)
+
 // Network describes the current network configuration in terms of nodes and
 // processes.
 type Network struct {
 	grpcClients map[string]GrpcClient
 	nodes       []*DuskNode
 	processes   []*os.Process
+
+	NetworkType byte
 }
 
 // AddNode to the network.
@@ -146,14 +155,16 @@ func (n *Network) Bootstrap(workspace string) error {
 		return err
 	}
 
-	// Start voucher seeder.
-	if len(seederExec) > 0 {
-		if err := n.start(workspace, seederExec); err != nil {
-			return err
+	// Start voucher seeder if gossip network is selected.
+	if n.NetworkType == GossipNetwork {
+		if len(seederExec) > 0 {
+			if err := n.start(workspace, seederExec); err != nil {
+				return err
+			}
+		} else {
+			// If path not provided, then it's assumed that the seeder is already running.
+			log.Warnf("Seeder path not provided. Please, ensure dusk-seeder is already running")
 		}
-	} else {
-		// If path not provided, then it's assumed that the seeder is already running.
-		log.Warnf("Seeder path not provided. Please, ensure dusk-seeder is already running")
 	}
 
 	if MOCK_ADDRESS != "" {

--- a/harness/tests/localnet_test.go
+++ b/harness/tests/localnet_test.go
@@ -66,7 +66,7 @@ func TestMain(m *testing.M) {
 
 	// Select network type based on nodes profile name
 	switch tomlProfile {
-	case "":
+	case "", "default":
 		{
 			tomlProfile = "default"
 			localNet.NetworkType = engine.GossipNetwork

--- a/harness/tests/localnet_test.go
+++ b/harness/tests/localnet_test.go
@@ -288,17 +288,6 @@ func TestMeasureNetworkTPS(t *testing.T) {
 
 	defaultLocktime := uint64(100000)
 
-	// Send a bunch of Bid transactions
-	// All nodes are Block Generators
-	for i := uint(0); i < uint(localNet.Size()); i++ {
-		time.Sleep(100 * time.Millisecond)
-		logrus.Infof("Node %d sending a Bid transaction", i)
-
-		if _, err := localNet.SendBidCmd(i, 100, defaultLocktime); err != nil {
-			logrus.Error(err.Error())
-		}
-	}
-
 	// Send a bunch of Stake transactions
 	// All nodes are Provisioners
 	for i := uint(0); i < uint(localNet.Size()); i++ {

--- a/harness/tests/localnet_test.go
+++ b/harness/tests/localnet_test.go
@@ -64,8 +64,17 @@ func TestMain(m *testing.M) {
 		localNetSize = currentLocalNetSize
 	}
 
-	if tomlProfile == "" {
-		tomlProfile = "default"
+	// Select network type based on nodes profile name
+	switch tomlProfile {
+	case "":
+		{
+			tomlProfile = "default"
+			localNet.NetworkType = engine.GossipNetwork
+		}
+	case "kadcast":
+		{
+			localNet.NetworkType = engine.KadcastNetwork
+		}
 	}
 
 	fmt.Println("GRPC Session enabled?", localNet.IsSessionRequired())

--- a/pkg/config/consts.go
+++ b/pkg/config/consts.go
@@ -36,6 +36,10 @@ const (
 	KadcastInitialHeight byte = 128
 )
 
+var (
+	KadcastInitHeader = []byte{KadcastInitialHeight}
+)
+
 // DecodeGenesis marshals a genesis block into a buffer.
 func DecodeGenesis() *block.Block {
 	cfg, err := genesis.GetPresetConfig(Get().General.Network)

--- a/pkg/config/consts.go
+++ b/pkg/config/consts.go
@@ -36,9 +36,8 @@ const (
 	KadcastInitialHeight byte = 128
 )
 
-var (
-	KadcastInitHeader = []byte{KadcastInitialHeight}
-)
+// KadcastInitHeader is used as default initial kadcast message header.
+var KadcastInitHeader = []byte{KadcastInitialHeight}
 
 // DecodeGenesis marshals a genesis block into a buffer.
 func DecodeGenesis() *block.Block {

--- a/pkg/core/candidate/requestor.go
+++ b/pkg/core/candidate/requestor.go
@@ -13,6 +13,7 @@ import (
 	"errors"
 	"sync"
 
+	"github.com/dusk-network/dusk-blockchain/pkg/config"
 	"github.com/dusk-network/dusk-blockchain/pkg/core/data/block"
 	"github.com/dusk-network/dusk-blockchain/pkg/p2p/wire/message"
 	"github.com/dusk-network/dusk-blockchain/pkg/p2p/wire/topics"
@@ -85,7 +86,14 @@ func (r *Requestor) publishGetCandidate(hash []byte) error {
 		return err
 	}
 
-	r.publisher.Publish(topics.Gossip, message.New(topics.GetCandidate, *buf))
+	if config.Get().Kadcast.Enabled {
+		m := message.NewWithHeader(topics.GetCandidate, *buf, config.KadcastInitHeader)
+		r.publisher.Publish(topics.Kadcast, m)
+	} else {
+		m := message.New(topics.GetCandidate, *buf)
+		r.publisher.Publish(topics.Gossip, m)
+	}
+
 	return nil
 }
 

--- a/pkg/core/candidate/requestor.go
+++ b/pkg/core/candidate/requestor.go
@@ -86,13 +86,14 @@ func (r *Requestor) publishGetCandidate(hash []byte) error {
 		return err
 	}
 
+	m := message.NewWithHeader(topics.GetCandidate, *buf, config.KadcastInitHeader)
+
 	if config.Get().Kadcast.Enabled {
-		m := message.NewWithHeader(topics.GetCandidate, *buf, config.KadcastInitHeader)
 		r.publisher.Publish(topics.Kadcast, m)
-	} else {
-		m := message.New(topics.GetCandidate, *buf)
-		r.publisher.Publish(topics.Gossip, m)
+		return nil
 	}
+
+	r.publisher.Publish(topics.Gossip, m)
 
 	return nil
 }

--- a/pkg/core/consensus/agreement/agreement_test.go
+++ b/pkg/core/consensus/agreement/agreement_test.go
@@ -109,7 +109,7 @@ func TestRequestor(t *testing.T) {
 		m := <-c
 		b := m.Payload().(message.SafeBuffer).Buffer
 
-		request, err := message.Unmarshal(&b)
+		request, err := message.Unmarshal(&b, nil)
 		assert.NoError(t, err)
 
 		if request.Category() == topics.GetCandidate {

--- a/pkg/core/consensus/agreement/step.go
+++ b/pkg/core/consensus/agreement/step.go
@@ -64,15 +64,14 @@ func (s *Loop) Run(ctx context.Context, roundQueue *consensus.Queue, agreementCh
 	evs := roundQueue.Flush(r.Round)
 
 	for _, ev := range evs {
-		go collectEvent(h, acc, ev.Payload().(message.Agreement), config.KadcastInitHeader, s.Emitter)
+		go collectEvent(h, acc, ev, s.Emitter)
 	}
 
 	for {
 		select {
 		case m := <-agreementChan:
 			if s.shouldCollectNow(m, r.Round, roundQueue) {
-				msg := m.Payload().(message.Agreement)
-				go collectEvent(h, acc, msg, m.Header(), s.Emitter)
+				go collectEvent(h, acc, m, s.Emitter)
 			}
 		case evs := <-acc.CollectedVotesChan:
 			lg.
@@ -147,14 +146,19 @@ func (s *Loop) requestCandidate(ctx context.Context, hash []byte) (block.Block, 
 	return s.requestor.RequestCandidate(ctx, hash)
 }
 
-func collectEvent(h *handler, accumulator *Accumulator, a message.Agreement, msgHeader []byte, e *consensus.Emitter) {
+func collectEvent(h *handler, accumulator *Accumulator, ev message.Message, e *consensus.Emitter) {
+	a := ev.Payload().(message.Agreement)
+
 	hdr := a.State()
 	if !h.IsMember(hdr.PubKeyBLS, hdr.Round, hdr.Step) {
 		return
 	}
 
+	// TODO: As an optimization, ev.Header() should be used instead of always starting from KadcastInitHeader.
+	m := message.NewWithHeader(topics.Agreement, a.Copy().(message.Agreement), config.KadcastInitHeader)
+
 	// Once the event is verified, we can republish it.
-	if err := e.Republish(message.New(topics.Agreement, a.Copy().(message.Agreement)), msgHeader); err != nil {
+	if err := e.Republish(m); err != nil {
 		lg.WithError(err).Error("could not republish agreement event")
 	}
 

--- a/pkg/core/consensus/agreement/step.go
+++ b/pkg/core/consensus/agreement/step.go
@@ -10,6 +10,7 @@ import (
 	"context"
 	"time"
 
+	"github.com/dusk-network/dusk-blockchain/pkg/config"
 	"github.com/dusk-network/dusk-blockchain/pkg/core/candidate"
 	"github.com/dusk-network/dusk-blockchain/pkg/core/consensus"
 	"github.com/dusk-network/dusk-blockchain/pkg/core/data/block"
@@ -61,8 +62,9 @@ func (s *Loop) Run(ctx context.Context, roundQueue *consensus.Queue, agreementCh
 	}()
 
 	evs := roundQueue.Flush(r.Round)
+
 	for _, ev := range evs {
-		go collectEvent(h, acc, ev.Payload().(message.Agreement), s.Emitter)
+		go collectEvent(h, acc, ev.Payload().(message.Agreement), config.KadcastInitHeader, s.Emitter)
 	}
 
 	for {
@@ -70,7 +72,7 @@ func (s *Loop) Run(ctx context.Context, roundQueue *consensus.Queue, agreementCh
 		case m := <-agreementChan:
 			if s.shouldCollectNow(m, r.Round, roundQueue) {
 				msg := m.Payload().(message.Agreement)
-				go collectEvent(h, acc, msg, s.Emitter)
+				go collectEvent(h, acc, msg, m.Header(), s.Emitter)
 			}
 		case evs := <-acc.CollectedVotesChan:
 			lg.
@@ -145,14 +147,14 @@ func (s *Loop) requestCandidate(ctx context.Context, hash []byte) (block.Block, 
 	return s.requestor.RequestCandidate(ctx, hash)
 }
 
-func collectEvent(h *handler, accumulator *Accumulator, a message.Agreement, e *consensus.Emitter) {
+func collectEvent(h *handler, accumulator *Accumulator, a message.Agreement, msgHeader []byte, e *consensus.Emitter) {
 	hdr := a.State()
 	if !h.IsMember(hdr.PubKeyBLS, hdr.Round, hdr.Step) {
 		return
 	}
 
 	// Once the event is verified, we can republish it.
-	if err := e.Gossip(message.New(topics.Agreement, a.Copy().(message.Agreement))); err != nil {
+	if err := e.Republish(message.New(topics.Agreement, a.Copy().(message.Agreement)), msgHeader); err != nil {
 		lg.WithError(err).Error("could not republish agreement event")
 	}
 

--- a/pkg/core/consensus/comms.go
+++ b/pkg/core/consensus/comms.go
@@ -8,7 +8,6 @@ package consensus
 
 import (
 	"bytes"
-	"errors"
 	"time"
 
 	"github.com/dusk-network/dusk-blockchain/pkg/config"
@@ -135,26 +134,21 @@ func (e *Emitter) Gossip(msg message.Message) error {
 }
 
 // Kadcast propagates a message in Kadcast network.
-func (e *Emitter) Kadcast(msg message.Message, h byte) error {
+func (e *Emitter) Kadcast(msg message.Message) error {
 	buf, err := message.Marshal(msg)
 	if err != nil {
 		return err
 	}
 
-	serialized := message.NewWithHeader(msg.Category(), buf, []byte{h})
+	serialized := message.NewWithHeader(msg.Category(), buf, msg.Header())
 	e.EventBus.Publish(topics.Kadcast, serialized)
 	return nil
 }
 
 // Republish reroutes message propagation to either Gossip or Kadcast network.
-func (e *Emitter) Republish(msg message.Message, header []byte) error {
+func (e *Emitter) Republish(msg message.Message) error {
 	if config.Get().Kadcast.Enabled {
-		if len(header) > 0 && header[0] <= config.KadcastInitialHeight {
-			h := header[0]
-			return e.Kadcast(msg, h)
-		}
-
-		return errors.New("unknown kadcast height")
+		return e.Kadcast(msg)
 	}
 
 	return e.Gossip(msg)

--- a/pkg/core/consensus/reduction/firststep/step.go
+++ b/pkg/core/consensus/reduction/firststep/step.go
@@ -159,8 +159,10 @@ func (p *Phase) collectReduction(ctx context.Context, r message.Reduction, round
 		return nil
 	}
 
+	m := message.NewWithHeader(topics.Reduction, r, config.KadcastInitHeader)
+
 	// Once the event is verified, we can republish it.
-	if err := p.Emitter.Republish(message.New(topics.Reduction, r), config.KadcastInitHeader); err != nil {
+	if err := p.Emitter.Republish(m); err != nil {
 		lg.WithError(err).Error("could not republish reduction event")
 	}
 

--- a/pkg/core/consensus/reduction/firststep/step.go
+++ b/pkg/core/consensus/reduction/firststep/step.go
@@ -11,6 +11,7 @@ import (
 	"context"
 	"time"
 
+	"github.com/dusk-network/dusk-blockchain/pkg/config"
 	"github.com/dusk-network/dusk-blockchain/pkg/core/candidate"
 	"github.com/dusk-network/dusk-blockchain/pkg/core/consensus"
 	"github.com/dusk-network/dusk-blockchain/pkg/core/consensus/header"
@@ -159,7 +160,7 @@ func (p *Phase) collectReduction(ctx context.Context, r message.Reduction, round
 	}
 
 	// Once the event is verified, we can republish it.
-	if err := p.Emitter.Gossip(message.New(topics.Reduction, r)); err != nil {
+	if err := p.Emitter.Republish(message.New(topics.Reduction, r), config.KadcastInitHeader); err != nil {
 		lg.WithError(err).Error("could not republish reduction event")
 	}
 

--- a/pkg/core/consensus/reduction/reduction.go
+++ b/pkg/core/consensus/reduction/reduction.go
@@ -82,7 +82,8 @@ func (r *Reduction) SendReduction(round uint64, step uint8, hash []byte) {
 	red := message.NewReduction(hdr)
 	red.SignedHash = sig
 
-	if err := r.Republish(message.New(topics.Reduction, *red), config.KadcastInitHeader); err != nil {
+	m := message.NewWithHeader(topics.Reduction, *red, config.KadcastInitHeader)
+	if err := r.Republish(m); err != nil {
 		panic(err)
 	}
 }

--- a/pkg/core/consensus/reduction/reduction.go
+++ b/pkg/core/consensus/reduction/reduction.go
@@ -9,6 +9,7 @@ package reduction
 import (
 	"time"
 
+	"github.com/dusk-network/dusk-blockchain/pkg/config"
 	"github.com/dusk-network/dusk-blockchain/pkg/core/consensus"
 	"github.com/dusk-network/dusk-blockchain/pkg/core/consensus/header"
 	"github.com/dusk-network/dusk-blockchain/pkg/p2p/wire/message"
@@ -81,7 +82,7 @@ func (r *Reduction) SendReduction(round uint64, step uint8, hash []byte) {
 	red := message.NewReduction(hdr)
 	red.SignedHash = sig
 
-	if err := r.Gossip(message.New(topics.Reduction, *red)); err != nil {
+	if err := r.Republish(message.New(topics.Reduction, *red), config.KadcastInitHeader); err != nil {
 		panic(err)
 	}
 }

--- a/pkg/core/consensus/reduction/secondstep/step.go
+++ b/pkg/core/consensus/reduction/secondstep/step.go
@@ -12,6 +12,7 @@ import (
 	"encoding/hex"
 	"time"
 
+	"github.com/dusk-network/dusk-blockchain/pkg/config"
 	"github.com/dusk-network/dusk-blockchain/pkg/core/consensus"
 	"github.com/dusk-network/dusk-blockchain/pkg/core/consensus/header"
 	"github.com/dusk-network/dusk-blockchain/pkg/core/consensus/reduction"
@@ -92,6 +93,7 @@ func (p *Phase) Run(ctx context.Context, queue *consensus.Queue, evChan chan mes
 	for _, ev := range queue.GetEvents(r.Round, step) {
 		if ev.Category() == topics.Reduction {
 			rMsg := ev.Payload().(message.Reduction)
+
 			if !p.handler.IsMember(rMsg.Sender(), r.Round, step) {
 				continue
 			}
@@ -169,7 +171,7 @@ func (p *Phase) collectReduction(r message.Reduction, round uint64, step uint8) 
 	}
 
 	// Once the event is verified, we can republish it.
-	if err := p.Emitter.Gossip(message.New(topics.Reduction, r)); err != nil {
+	if err := p.Emitter.Republish(message.New(topics.Reduction, r.Copy().(message.Reduction)), config.KadcastInitHeader); err != nil {
 		lg.WithError(err).Error("could not republish reduction event")
 	}
 
@@ -249,7 +251,7 @@ func (p *Phase) sendAgreement(round uint64, step uint8, svm *message.StepVotesMs
 		"step":  step,
 	}).Traceln("gossiping_agreement")
 
-	if err := p.Gossip(message.New(topics.Agreement, *ev)); err != nil {
+	if err := p.Republish(message.New(topics.Agreement, *ev), config.KadcastInitHeader); err != nil {
 		lg.
 			WithError(err).
 			Error("gossip the agreement reported error, consensus is continuing but this needs to be investigated!")

--- a/pkg/core/consensus/reduction/secondstep/step.go
+++ b/pkg/core/consensus/reduction/secondstep/step.go
@@ -170,8 +170,10 @@ func (p *Phase) collectReduction(r message.Reduction, round uint64, step uint8) 
 		return nil
 	}
 
+	m := message.NewWithHeader(topics.Reduction, r.Copy().(message.Reduction), config.KadcastInitHeader)
+
 	// Once the event is verified, we can republish it.
-	if err := p.Emitter.Republish(message.New(topics.Reduction, r.Copy().(message.Reduction)), config.KadcastInitHeader); err != nil {
+	if err := p.Emitter.Republish(m); err != nil {
 		lg.WithError(err).Error("could not republish reduction event")
 	}
 
@@ -251,7 +253,8 @@ func (p *Phase) sendAgreement(round uint64, step uint8, svm *message.StepVotesMs
 		"step":  step,
 	}).Traceln("gossiping_agreement")
 
-	if err := p.Republish(message.New(topics.Agreement, *ev), config.KadcastInitHeader); err != nil {
+	m := message.NewWithHeader(topics.Agreement, *ev, config.KadcastInitHeader)
+	if err := p.Republish(m); err != nil {
 		lg.
 			WithError(err).
 			Error("gossip the agreement reported error, consensus is continuing but this needs to be investigated!")

--- a/pkg/core/consensus/selection/step.go
+++ b/pkg/core/consensus/selection/step.go
@@ -169,7 +169,8 @@ func (p *Phase) collectScore(sc message.Score, msgHeader []byte) error {
 
 	// Once the event is verified, and has passed all preliminary checks,
 	// we can republish it to the network.
-	if err := p.Republish(message.New(topics.Score, sc), msgHeader); err != nil {
+	m := message.NewWithHeader(topics.Score, sc, msgHeader)
+	if err := p.Republish(m); err != nil {
 		lg.WithError(err).WithField("kadcast_enabled", config.Get().Kadcast.Enabled).
 			Error("could not republish score event")
 	}

--- a/pkg/core/consensus/testing/utils.go
+++ b/pkg/core/consensus/testing/utils.go
@@ -72,7 +72,7 @@ func (g *gossipRouter) route(m message.Message) {
 	if g.d.HasAnywhere(&b) {
 		// The incoming message will be a message.SafeBuffer, as it's coming from
 		// the consensus.Emitter.
-		m, err := message.Unmarshal(&b)
+		m, err := message.Unmarshal(&b, nil)
 		if err != nil {
 			panic(err)
 		}

--- a/pkg/p2p/kadcast/writer.go
+++ b/pkg/p2p/kadcast/writer.go
@@ -102,6 +102,10 @@ func (w *Writer) Serve() {
 			return 2
 		case topics.Score:
 			return 3
+		case topics.Reduction:
+			return 4
+		case topics.Agreement:
+			return 5
 		}
 
 		return 1

--- a/pkg/p2p/kadcast/writer.go
+++ b/pkg/p2p/kadcast/writer.go
@@ -118,7 +118,6 @@ func (w *Writer) Serve() {
 
 	l2 := eventbus.NewStreamListenerWithParams(w, MaxWriterQueueSize, priorityMapper)
 	w.kadcastPointSubscription = w.subscriber.Subscribe(topics.KadcastPoint, l2)
-
 }
 
 func (w *Writer) Write(data, header []byte, priority byte) (int, error) {

--- a/pkg/p2p/kadcast/writer.go
+++ b/pkg/p2p/kadcast/writer.go
@@ -106,6 +106,8 @@ func (w *Writer) Serve() {
 			return 4
 		case topics.Agreement:
 			return 5
+		case topics.Candidate, topics.GetCandidate:
+			return 6
 		}
 
 		return 1
@@ -116,6 +118,7 @@ func (w *Writer) Serve() {
 
 	l2 := eventbus.NewStreamListenerWithParams(w, MaxWriterQueueSize, priorityMapper)
 	w.kadcastPointSubscription = w.subscriber.Subscribe(topics.KadcastPoint, l2)
+
 }
 
 func (w *Writer) Write(data, header []byte, priority byte) (int, error) {

--- a/pkg/p2p/peer/peer.go
+++ b/pkg/p2p/peer/peer.go
@@ -326,14 +326,10 @@ func (p *Reader) ReadLoop(ctx context.Context, ringBuf *ring.Buffer) {
 		go func() {
 			// TODO: error here should be checked in order to decrease reputation
 			// or blacklist spammers
-			startTime := time.Now().UnixNano()
-
 			if _, err = p.processor.Collect(p.Addr(), message, ringBuf, p.services, nil); err != nil {
 				l.WithField("process", "readloop").WithField("cs", hex.EncodeToString(cs)).
 					WithError(err).Error("failed to process message")
 			}
-
-			p.logWireMsg(startTime, cs, message)
 		}()
 
 		// Reset the keepalive timer
@@ -352,23 +348,6 @@ func (p *Reader) keepAliveLoop(ctx context.Context, timer *time.Timer) {
 			timer.Stop()
 			return
 		}
-	}
-}
-
-func (p *Reader) logWireMsg(startTime int64, cs, msg []byte) {
-	if l.Logger.GetLevel() == log.TraceLevel {
-		duration := float64(time.Now().UnixNano()-startTime) / 1000000
-
-		var topicName string
-		if len(msg) > 0 {
-			topicName = topics.Topic(msg[0]).String()
-		}
-
-		l.WithField("process", "readloop").WithField("cs", hex.EncodeToString(cs)).
-			WithField("len", len(msg)).
-			WithField("ms", duration).
-			WithField("topic", topicName).
-			Trace("gossip message")
 	}
 }
 

--- a/pkg/p2p/peer/processor.go
+++ b/pkg/p2p/peer/processor.go
@@ -10,6 +10,7 @@ import (
 	"bytes"
 	"errors"
 	"fmt"
+	"time"
 
 	log "github.com/sirupsen/logrus"
 
@@ -53,19 +54,34 @@ func (m *MessageProcessor) Collect(srcPeerID string, packet []byte, respRingBuf 
 		return nil, errors.New("empty packet provided")
 	}
 
+	defer m.trace(time.Now().UnixNano(), packet)
+
 	b := bytes.NewBuffer(packet)
 	topic := topics.Topic(b.Bytes()[0])
 
-	msg, err := message.Unmarshal(b)
+	msg, err := message.Unmarshal(b, header)
 	if err != nil {
 		return nil, fmt.Errorf("error while unmarshaling: %s - topic: %s", err, topic)
 	}
 
-	if header != nil {
-		msg = message.NewWithHeader(msg.Category(), msg.Payload(), header)
-	}
-
 	return m.process(srcPeerID, msg, respRingBuf, services)
+}
+
+func (m *MessageProcessor) trace(st int64, msg []byte) {
+	if l.Logger.GetLevel() == log.TraceLevel {
+		duration := float64(time.Now().UnixNano()-st) / 1000000
+
+		var topicName string
+		if len(msg) > 0 {
+			topicName = topics.Topic(msg[0]).String()
+		}
+
+		l.WithField("process", "msg_processor").
+			WithField("len", len(msg)).
+			WithField("ms", duration).
+			WithField("topic", topicName).
+			Trace("wire message")
+	}
 }
 
 func (m *MessageProcessor) shouldBeCached(t topics.Topic) bool {

--- a/pkg/p2p/wire/message/message.go
+++ b/pkg/p2p/wire/message/message.go
@@ -97,6 +97,7 @@ func Clone(m Message) (Message, error) {
 		category:  m.Category(),
 		marshaled: &b,
 		payload:   m.Payload().Copy(),
+		header:    m.Header(),
 	}, nil
 }
 
@@ -229,10 +230,10 @@ func (m *simple) initPayloadBuffer(b bytes.Buffer) {
 // Unmarshal mutates the buffer by extracting the topic. It create the Message
 // by setting the topic and unmarshaling the payload into the proper structure
 // It also caches the serialized form within the message.
-func Unmarshal(b *bytes.Buffer) (Message, error) {
+func Unmarshal(b *bytes.Buffer, h []byte) (Message, error) {
 	var err error
 
-	msg := &simple{}
+	msg := &simple{header: h}
 	msg.initPayloadBuffer(*b)
 
 	topic, err := topics.Extract(b)

--- a/pkg/util/nativeutils/eventbus/mock.go
+++ b/pkg/util/nativeutils/eventbus/mock.go
@@ -264,7 +264,7 @@ func (r *RouterStreamer) Write(p []byte) (n int, err error) {
 	for i := 0; i < len(r.peers); i++ {
 		b := bytes.NewBuffer(p)
 
-		msg, err := message.Unmarshal(b)
+		msg, err := message.Unmarshal(b, nil)
 		if err != nil {
 			return 0, err
 		}

